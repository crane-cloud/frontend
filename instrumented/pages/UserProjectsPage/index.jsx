import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import Modal from "../../components/Modal";
import styles from "./UserProjectsPage.module.css";
import InformationBar from "../../components/InformationBar";
import { ReactComponent as ButtonPlus } from "../../assets/images/buttonplus.svg";
import Header from "../../components/Header";
import getClustersList from "../../redux/actions/clusters";
import CreateProject from "../../components/CreateProject";
import getUserProjects from "../../redux/actions/projectsList";
import getUserCredits from "../../redux/actions/userCredits";
import ProjectCard from "../../components/ProjectCard";
import PrimaryButton from "../../components/PrimaryButton";
import Spinner from "../../components/Spinner";
import { handlePatchRequest } from "../../apis/apis.js";
import { ReactComponent as User } from "../../assets/images/user.svg";
import { ReactComponent as Users } from "../../assets/images/users.svg";
import "../../index.css";

class UserProjectsPage extends React.Component {
  constructor(props) {
    super(props);
    this.initialState = {
      openCreateComponent: false,
      Searchword: "",
      SearchList: [],
      myProjectsList: [],
      sharedProjectsList: [],
      showInviteModel: false,
      selectedProjects: "My projects",
      inviteeProjectId: "",
      inviteeModelRole: "",
      decliningInvitation: false,
      acceptingInvitation: false,
      invitationError: "",
      currentTab: "My Projects",
    };
    this.state = this.initialState;
    this.openProjectCreateComponent =
      this.openProjectCreateComponent.bind(this);
    this.callbackProjectCreateComponent =
      this.callbackProjectCreateComponent.bind(this);
    this.searchThroughProjects = this.searchThroughProjects.bind(this);
    this.handleCallbackSearchword = this.handleCallbackSearchword.bind(this);
    this.showInvitationModel = this.showInvitationModel.bind(this);
    this.hideInvitationModel = this.hideInvitationModel.bind(this);
    this.handleInvitationAcceptence =
      this.handleInvitationAcceptence.bind(this);
    this.handleInvitationDecline = this.handleInvitationDecline.bind(this);
    this.filterProjects = this.filterProjects.bind(this);
  }

  componentDidMount() {
    const {
      getClustersList,
      getUserProjects,
      data,
      getUserCredits,
    } = this.props;
    getUserProjects(data.id);
    getClustersList();
    getUserCredits(data.id);
  }

  componentDidUpdate(prevProps, prevState) {
    const {
      isAdded,
      getClustersList,
      getUserProjects,
      isDeleted,
      isUpdated,
      clearUpdateProjectState,
      isFetched,
    } = this.props;
    const { Searchword } = this.state;

    if (isDeleted !== prevProps.isDeleted) {
      getUserProjects();
      getClustersList();
    }

    if (isUpdated !== prevProps.isUpdated) {
      clearUpdateProjectState();
      getUserProjects();
      getClustersList();
    }

    if (isAdded !== prevProps.isAdded) {
      getUserProjects();
      this.setState(this.initialState);
    }
    if (isFetched !== prevProps.isFetched) {
      //call filter
      const filteredData = this.filterProjects();
      this.setState({
        myProjectsList: filteredData.myProjects,
        sharedProjectsList: filteredData.sharedProjects,
      });
    }
    if (Searchword !== prevState.Searchword) {
      this.searchThroughProjects();
    }
  }

  openProjectCreateComponent() {
    this.setState({ openCreateComponent: true });
  }
  callbackProjectCreateComponent() {
    const {
      getUserProjects
    } = this.props;
    getUserProjects()
    this.setState(this.initialState);
  }
  searchThroughProjects() {
    const { Searchword } = this.state;
    const { projects } = this.props;
    let searchResult = [];
    projects.forEach((element) => {
      if (element.name.toLowerCase().includes(Searchword.toLowerCase())) {
        searchResult.push(element);
      }
    });
    this.setState({
      SearchList: searchResult.sort((a, b) =>
        b.date_created > a.date_created ? 1 : -1
      ),
    });
  }
  hideInvitationModel() {
    this.setState({ showInviteModel: false });
  }
  showInvitationModel(inviteeProjectId, inviteeModelRole) {
    this.setState({
      showInviteModel: true,
      inviteeProjectId: inviteeProjectId,
      inviteeModelRole: inviteeModelRole,
    });
  }
  handleCallbackSearchword(word) {
    this.setState({
      Searchword: word,
    });
  }
  handleInvitationDecline() {
    const { getUserProjects } = this.props;
    this.setState({
      decliningInvitation: true,
    });
    const { inviteeProjectId } = this.state;
    handlePatchRequest(`/projects/${inviteeProjectId}/users/handle_invite`, {
      accepted_collaboration_invite: false,
    })
      .then((response) => {
        this.setState({
          decliningInvitation: false,
        });
        this.hideInvitationModel();
        getUserProjects();
      })
      .catch((error) => {
        this.setState({
          invitationError: "Something went wrong",
          decliningInvitation: false,
        });
        this.hideInvitationModel();
      });
  }
  handleInvitationAcceptence() {
    const { getUserProjects } = this.props;
    this.setState({
      acceptingInvitation: true,
    });
    const { inviteeProjectId } = this.state;

    handlePatchRequest(`/projects/${inviteeProjectId}/users/handle_invite`, {
      accepted_collaboration_invite: true,
    })
      .then((response) => {
        this.setState({
          acceptingInvitation: false,
        });
        this.hideInvitationModel();
        getUserProjects();
      })
      .catch((error) => {
        this.setState({
          invitationError: "Something went wrong",
          acceptingInvitation: false,
        });
        this.hideInvitationModel();
      });
  }
  filterProjects() {
    const { projects, data } = this.props;
    let myProjects = [];
    let sharedProjects = [];
    projects.forEach((element) => {
      if (element.owner_id === data.id) {
        myProjects.push(element);
      } else {
        sharedProjects.push(element);
      }
    });
    return {
      sharedProjects,
      myProjects,
    };
  }

  render() {
    const {
      openCreateComponent,
      Searchword,
      SearchList,
      showInviteModel,
      selectedProjects,
      inviteeModelRole,
      decliningInvitation,
      acceptingInvitation,
      invitationError,
      sharedProjectsList,
      myProjectsList,
      currentTab,
    } = this.state;
    const {
      projects,
      isRetrieving,
      isFetched,
      match: { params },
      credits,
      data,
    } = this.props;
    const displayProjects =
      selectedProjects === "My projects" ? myProjectsList : sharedProjectsList;
    const sortedProjects = displayProjects.sort((a, b) =>
      b.date_created > a.date_created ? 1 : -1
    );
    return (
      <div className={styles.Page}>
        {openCreateComponent ? (
          <CreateProject
            closeComponent={this.callbackProjectCreateComponent}
            params={params}
          />
        ) : (
          <div>
            <div className={styles.TopRow}>
              <Header credits={credits?.amount} />
              <InformationBar
                header="Projects"
                showBtn
                buttontext="+ New Project"
                showSearchBar
                placeholder="Search through projects"
                btnAction={this.openProjectCreateComponent}
                searchAction={this.handleCallbackSearchword}
              />
            </div>
            <div className={styles.MainRow}>
              <div className={`${styles.SelectProjects} SmallContainer`}>
                <div className={styles.ProjectCategoryHeadings}>
                  <span
                    className={
                      currentTab === "My Projects"
                        ? styles.CurrentTab
                        : styles.Tab
                    }
                    onClick={() => {
                      this.setState({
                        selectedProjects: "My projects",
                        currentTab: "My Projects",
                      });
                    }}
                  >
                    <div className={styles.ProjectCategories}>
                      <span>
                        <User className={styles.SmallerIcon} />
                      </span>
                      <span>My Projects</span>
                    </div>
                  </span>
                  <span
                    className={
                      currentTab === "Shared Projects"
                        ? styles.CurrentTab
                        : styles.Tab
                    }
                    onClick={() => {
                      this.setState({
                        selectedProjects: "Shared Projects",
                        currentTab: "Shared Projects",
                      });
                    }}
                  >
                    <div className={styles.ProjectCategories}>
                      <span>
                        <Users className={styles.SmallerIcon} />
                      </span>
                      <span>Shared Projects</span>
                    </div>
                  </span>
                </div>
              </div>
              {isRetrieving ? (
                <div className={styles.NoResourcesMessage}>
                  <div className={styles.SpinnerWrapper}>
                    <Spinner size="big" />
                  </div>
                </div>
              ) : Searchword !== "" ? (
                <div className={`${styles.ProjectList}  SmallContainer`}>
                  {isFetched &&
                    SearchList !== undefined &&
                    SearchList.map((project) => (
                      <ProjectCard
                        key={project.id}
                        name={project.name}
                        description={project.description}
                        cardID={project.id}
                        acceptInviteCallBackModel={this.showInvitationModel}
                        userID={data.id}
                        ownerId={project.owner_id}
                        apps_count={project.apps_count}
                      />
                    ))}
                </div>
              ) : (
                <div className={`${styles.ProjectList}  SmallContainer`}>
                  {isFetched &&
                    sortedProjects !== undefined &&
                    sortedProjects.map((project) => (
                      <ProjectCard
                        key={project.id}
                        name={project.name}
                        description={project.description}
                        cardID={project.id}
                        userID={data.id}
                        ownerId={project.owner_id}
                        acceptInviteCallBackModel={this.showInvitationModel}
                        apps_count={project.apps_count}
                      />
                    ))}
                </div>
              )}
              {showInviteModel === true && (
                <div className={styles.ProjectDeleteModel}>
                  <Modal
                    showModal={showInviteModel}
                    onClickAway={this.hideInvitationModel}
                  >
                    <div className={styles.ModelContent}>
                      <div className={styles.ModelHeader}>
                        Invitation to this project{" "}
                      </div>
                      <div className={styles.UpdateForm}>
                        <div className={styles.InformationText}>
                          You have been invited to join this project as (a/an)
                          <span className={styles.Role}>
                            {" "}
                            {inviteeModelRole}.
                          </span>
                        </div>
                        <div className={styles.InformationWarning}>
                          If you decline, you will not be able to see this
                          project again unless you are re-invited.
                        </div>
                        <div className={styles.UpdateProjectModelButtons}>
                          <PrimaryButton
                            color="red"
                            onClick={() => {
                              this.handleInvitationDecline();
                            }}
                          >
                            {decliningInvitation ? <Spinner /> : "Decline"}
                          </PrimaryButton>
                          <PrimaryButton
                            color="primary"
                            onClick={() => {
                              this.handleInvitationAcceptence();
                            }}
                          >
                            {acceptingInvitation ? <Spinner /> : "Accept"}
                          </PrimaryButton>
                        </div>
                        {invitationError && (
                          <div className={styles.InformationWarning}>
                            {invitationError}
                          </div>
                        )}
                      </div>
                    </div>
                  </Modal>
                </div>
              )}
              {displayProjects.length === 0 && (
                <div className={styles.NoResourcesMessage}>
                  {selectedProjects === "My projects" ? (
                    <div>
                      You haven’t created any projects yet. Click the &nbsp;{" "}
                      <ButtonPlus className={styles.ButtonPlusSmall} /> &nbsp;
                      button to add a project.
                    </div>
                  ) : (
                    <>No project has been shared with you as yet.</>
                  )}
                </div>
              )}
              {isFetched && projects.length === 0 && (
                <div className={styles.NoResourcesMessage}>
                  You haven’t created any projects yet. Click the &nbsp;{" "}
                  <ButtonPlus className={styles.ButtonPlusSmall} /> &nbsp;
                  button to add a project.
                </div>
              )}
              {!isRetrieving && !isFetched && (
                <div className={styles.NoResourcesMessage}>
                  Oops! Something went wrong! Failed to retrieve Projects.
                </div>
              )}
            </div>
          </div>
        )}
        <div className={styles.FooterRow}>
          <div>
            Copyright {new Date().getFullYear()} Crane Cloud. All Rights
            Reserved.
          </div>
        </div>
      </div>
    );
  }
}

UserProjectsPage.propTypes = {
  projects: PropTypes.arrayOf(PropTypes.shape({})),
  clusters: PropTypes.object,
  getClustersList: PropTypes.func.isRequired,
  getUserProjects: PropTypes.func.isRequired,
  data: PropTypes.shape({
    id: PropTypes.string,
  }).isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      userID: PropTypes.string,
    }).isRequired,
  }).isRequired,
  isFetched: PropTypes.bool,
  isRetrieving: PropTypes.bool,
};

UserProjectsPage.defaultProps = {
  clusters: [],
  projects: [],
  isFetched: false,
  isRetrieving: false,
};

export const mapStateToProps = (state) => {
  const { data } = state.user;

  const { clusters } = state.clustersReducer;
  const { isRetrieving, projects, isFetched } = state.userProjectsReducer;
  const { credits } = state.userCreditsReducer;
  return {
    data,
    isRetrieving,
    projects,
    clusters,
    isFetched,
    credits,
  };
};

const mapDispatchToProps = {
  getUserProjects,
  getClustersList,
  getUserCredits,
};

export default connect(mapStateToProps, mapDispatchToProps)(UserProjectsPage);
