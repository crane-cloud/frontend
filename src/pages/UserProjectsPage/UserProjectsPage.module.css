.ProjectList {
  display: flex;
  gap: 1rem;
  justify-content: flex-start;
  flex-wrap: wrap;
  padding: 2rem 0;
}

.ProjectFormErrorDiv {
  background-color: rgba(255, 255, 255, 0.5);
  color: #d63b3b;
  text-align: center;
  padding: 1rem 0;
  height: 1rem;
}

.ClusterDrop {
  width: 25rem;
  border: 2px solid #000000;
  box-sizing: border-box;
  font-size: inherit;
  padding: 1rem;
}

.ClusterNameOption {
  padding: 2rem 1.5rem;
  color: #000;
  background-color: #f9f9f9;
  font-weight: bold;
}

.MainRow > .ConnectModal {
  width: 100%;
  margin-left: 0;
}

.MainRow {
  background-color: #f1f1f1;
  min-height: calc(100vh - 14rem);
}

.NoResourcesMessage {
  flex-grow: 1;
  display: flex;
  font-size: 16px;
  text-align: center;
  justify-content: center;
  align-items: center;
  padding: 12rem;
}

.ButtonPlusSmall {
  width: 1rem;
  height: 1rem;
}

.FooterRow {
  padding: 1rem 3rem;
  background-color: #f1f1f1;
}

.SearchBar {
  flex-grow: 1;
  display: flex;
  padding: 2rem;
}


@media (max-width: 500px) {
  .ProjectList > div{
    width: 100%!important;
  }  
}
/* @media only screen and (max-width: 1100px) and (min-width: 901px) {
  .ProjectList {
    grid-template-columns: repeat(4, 1fr);
  }
}

@media only screen and (max-width: 900px) and (min-width: 481px) {
  .ProjectList {
    grid-template-columns: repeat(3, 1fr);
  }
}

} */
