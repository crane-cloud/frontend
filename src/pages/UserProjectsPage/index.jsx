import React, { useState, useEffect, useCallback } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import Modal from "../../components/Modal";
import styles from "./UserProjectsPage.module.css";
import InformationBar from "../../components/InformationBar";
import { ReactComponent as ButtonPlus } from "../../assets/images/buttonplus.svg";
import Header from "../../components/Header";
import Pagination from "../../components/Pagination";
import getClustersList from "../../redux/actions/clusters";
import CreateProject from "../../components/CreateProject";
import getUserProjects from "../../redux/actions/projectsList";
import getUserCredits from "../../redux/actions/userCredits";
import ProjectCard from "../../components/ProjectCard";
import PrimaryButton from "../../components/PrimaryButton";
import Spinner from "../../components/Spinner";
import { handlePatchRequest } from "../../apis/apis.js";
import "../../index.css";
import { Link } from "react-router-dom/cjs/react-router-dom.min.js";
import { useProjects } from "../../hooks/useProjects.js";

const UserProjectsPage = (props) => {
  const [openCreateComponent, setOpenCreateComponent] = useState(false);
  const [searchword, setSearchword] = useState("");
  // SearchList: [],
  const [projectsList, setProjectList] = useState([]);
  const [sharedProjectsList, setSharedProjectList] = useState([]);
  const [showInviteModel, setShowInviteModel] = useState(false);
  const [selectedProjects, setSelectedProjects] = useState("My projects");
  const [inviteeProjectId, setInviteeProjectId] = useState("");
  const [inviteeModelRole, setInviteeModelRole] = useState("");
  const [decliningInvitation, setDecliningInvitation] = useState(false);
  const [acceptingInvitation, setAcceptingInvitation] = useState(false);
  const [invitationError, setInvitationError] = useState("");
  const [currentTab, setCurrentTab] = useState("My Projects");
  const [currentPaginationPage, setCurrentPaginationPage] = useState(1);
  const [displayProjects, setDisplayProjects] = useState([]);
  const [userProjects, setUserProjects] = useState([]);

  const { data: projectsData, isLoading, isError, isFetched} =
    useProjects(currentPaginationPage, searchword);

  const {
    getUserProjects,
    getClustersList,
    getUserCredits,
    data,
    pagination,
    isRetrieving,
    credits,
    isAdded,
    isDeleted,
    isUpdated,
    clearUpdateProjectState,
    match: { params },
  } = props;


  const handleSharedProjectsTabChange = () => {
    setSelectedProjects("Shared Projects");
    setCurrentTab("Shared Projects");
  };

  const handleTabChange = useCallback(() => {
    setSelectedProjects("My projects");
    setCurrentTab("My Projects");
  }, [currentTab]);

  const handleTabAll = useCallback(
    (selectedOption) => {
      let displayProjects;
      if (selectedOption === "All") {
        displayProjects = projectsList.concat(sharedProjectsList);
      } else if (selectedOption === "My Projects") {
        displayProjects = projectsList;
      } else if (selectedOption === "Shared Projects") {
        displayProjects = sharedProjectsList;
      } else {
        displayProjects = projectsList.concat(sharedProjectsList);
      }
      setCurrentTab(selectedOption);
      setSelectedProjects(selectedOption);
      setDisplayProjects(displayProjects);
    },
    [projectsList, sharedProjectsList]
  );

  useEffect(() => {
    getClustersList();
    if(data.id){
    getUserCredits(data.id);
    }
  }, [data.id, currentPaginationPage]); 

  // useEffect(() => {
  //   if (isFetched) {
  //     const filteredData = filterProjects();
  //     setProjectList(filteredData.myProjects);
  //     setSharedProjectList(filteredData.sharedProjects);
  //     setDisplayProjects(
  //       filteredData.myProjects.concat(filteredData.sharedProjects)
  //     );
  //   }
  // },[isFetched]);

  // useEffect(() => {
  //   searchThroughProjects();
  // },[searchword]);

  useEffect(() => {
    if (isDeleted) {
      getUserProjects(currentPaginationPage);
      getClustersList();
    }

    if (isUpdated) {
      clearUpdateProjectState();
      getUserProjects(currentPaginationPage);
      getClustersList();
    }

    if (isAdded) {
      getUserProjects(currentPaginationPage);
      setUserProjects([]);
    }
  }, [
    isDeleted,
    isUpdated,
    isAdded,
    clearUpdateProjectState,
    getClustersList,
    currentPaginationPage,
    getUserProjects
  ]);

  const filterProjects = () => {
    let myProjects = [];
    let sharedProjects = [];

    projectsData.data?.data?.projects.forEach((element) => {
      if (element.owner_id === data.id) {
        myProjects.push(element);
      } else {
        sharedProjects.push(element);
      }
    });
    if (myProjects.length < sharedProjects.length) {
      setSelectedProjects("Shared Projects");
      setCurrentTab("Shared Projects");
    } else {
      setSelectedProjects("My projects");
      setCurrentTab("My projects");
    }
    return {myProjects, sharedProjects}
  }
  
  useEffect(() => {
    if (isFetched) {
      const filteredData = filterProjects();

      if (
        JSON.stringify(projectsList) !== JSON.stringify(filteredData.myProjects) ||
        JSON.stringify(sharedProjectsList) !== JSON.stringify(filteredData.sharedProjects)
      ) {
        setProjectList(filteredData.myProjects);
        setSharedProjectList(filteredData.sharedProjects);
        setDisplayProjects(filteredData.myProjects.concat(filteredData.sharedProjects))
    }
  }
  },[isFetched, projectsList, sharedProjectsList]);

  useEffect(() => {
    if (searchword) {
      searchThroughProjects();
    }
  },[searchword]);

  useEffect(() => {
    if (isFetched) {
      const combinedProjects = projectsList.concat(sharedProjectsList);
      setSelectedProjects("All");
      setCurrentTab("All");
      setDisplayProjects(combinedProjects);
    }
  }, [isFetched, projectsList, sharedProjectsList]);

  const openProjectCreateComponent = () => {
    setOpenCreateComponent(true);
  };
  const callbackProjectCreateComponent = useCallback(() => {
    setOpenCreateComponent(false);
  },[]);

  const searchThroughProjects = () => {
    setCurrentPaginationPage(1);
    getUserProjects(1, searchword);
  };

  const hideInvitationModel = () => {
    setShowInviteModel(false);
  };

  const showInvitationModel = (inviteeProjectId, inviteeModelRole) => {
    setShowInviteModel(true);
    setInviteeProjectId(inviteeProjectId);
    setInviteeModelRole(inviteeModelRole);
  };

  const handleInvitationDecline = () => {
    setDecliningInvitation(true);
    handlePatchRequest(`/projects/${inviteeProjectId}/users/handle_invite`, {
      accepted_collaboration_invite: false,
    })
      .then(() => {
        setDecliningInvitation(false);
        setShowInviteModel(false);
        getUserProjects(currentPaginationPage);
      })
      .catch(() => {
        setInvitationError("Something went wrong");
        setDecliningInvitation(false);
        setShowInviteModel(false);
      });
  };

  const handleInvitationAcceptence = () => {
    setAcceptingInvitation(true);
    handlePatchRequest(`/projects/${inviteeProjectId}/users/handle_invite`, {
      accepted_collaboration_invite: true,
    })
      .then(() => {
        setAcceptingInvitation(false);
        setShowInviteModel(false);
        getUserProjects(currentPaginationPage);
      })
      .catch(() => {
        setInvitationError("Something went wrong");
        setAcceptingInvitation(false);
        setShowInviteModel(false);
      });
  };
  
  const onPageChange = (page) => {
    setCurrentPaginationPage(page);
  };
  
  // const onFilterSelect = (selectedOption) => {
  //   setSelectedProjects(selectedOption);
  //   const updatedDisplayProjects =
  //     selectedOption === "All"
  //       ? projectsList.concat(sharedProjectsList)
  //       : selectedOption === "Shared Projects"
  //       ? sharedProjectsList
  //       : projectsList;

  //   setDisplayProjects(updatedDisplayProjects);
  // };

  return (
    <div className={styles.Page}>
      {openCreateComponent ? (
        <CreateProject
          closeComponent={callbackProjectCreateComponent}
          params={params}
        />
      ) : (
        <div>
          <div className={styles.TopRow}>
            <Header credits={credits?.amount} />
            <InformationBar
              header={
                <span>
                  <Link className="breadcrumb" to={`/dashboard`}>
                    Dashboard
                  </Link>
                  / Projects
                </span>
              }
              showBtn
              buttontext="+ New Project"
              showSearchBar
              selectedProjects={selectedProjects}
              myProjectsList={projectsList}
              sharedProjectsList={sharedProjectsList}
              viewFilter={true}
              placeholder="Search through projects"
              btnAction={openProjectCreateComponent}
              searchAction={setSearchword}
              adminRoute={data?.username === "admin"}
              // onFilterSelect={handleTabAll}
            />
          </div>
          <div className={styles.MainRow}>
            <div className={`${styles.SelectProjects} SmallContainer`}></div>
            {isLoading ? (
              <div className={styles.NoResourcesMessage}>
                <div className={styles.SpinnerWrapper}>
                  <Spinner size="big" />
                </div>
              </div>
            ) : searchword !== "" ? (
              <div className={`${styles.ProjectList}  SmallContainer`}>
                {isFetched &&
                  projectsData.data?.data?.projects!== undefined &&
                  projectsData.data?.data?.projects.map((project) => (
                    <ProjectCard
                      key={project.id}
                      name={project.name}
                      description={project.description}
                      cardID={project.id}
                      acceptInviteCallBackModel={showInvitationModel}
                      userID={data.id}
                      ownerId={project.owner_id}
                      apps_count={project.apps_count}
                      followers_count={project.followers_count}
                      disabled={project.disabled}
                      admin_disabled={project.admin_disabled}
                    />
                  ))}
              </div>
            ) : (
              <div className={`${styles.ProjectList}  SmallContainer`}>
                {isFetched &&
                  projectsData.data?.data?.projects!== undefined &&
                  projectsData.data?.data?.projects.map((project) => (
                    <ProjectCard
                      key={project.id}
                      name={project.name}
                      description={project.description}
                      cardID={project.id}
                      userID={data.id}
                      ownerId={project.owner_id}
                      acceptInviteCallBackModel={showInvitationModel}
                      followers_count={project.followers_count}
                      apps_count={project.apps_count}
                      disabled={project.disabled}
                      admin_disabled={project.admin_disabled}
                    />
                  ))}
              </div>
            )}

            {showInviteModel === true && (
              <div className={styles.ProjectDeleteModel}>
                <Modal
                  showModal={showInvitationModel}
                  onClickAway={hideInvitationModel}
                >
                  <div className={styles.ModelContent}>
                    <div className={styles.ModelHeader}>
                      Invitation to this project{" "}
                    </div>
                    <div className={styles.UpdateForm}>
                      <div className={styles.InformationText}>
                        You have been invited to join this project as (a/an)
                        <span className={styles.Role}>
                          {" "}
                          {inviteeModelRole}.
                        </span>
                      </div>
                      <div className={styles.InformationWarning}>
                        If you decline, you will not be able to see this project
                        again unless you are re-invited.
                      </div>
                      <div className={styles.UpdateProjectModelButtons}>
                        <PrimaryButton
                          color="red"
                          onClick={() => {
                            handleInvitationDecline();
                          }}
                        >
                          {decliningInvitation ? <Spinner /> : "Decline"}
                        </PrimaryButton>
                        <PrimaryButton
                          color="primary"
                          onClick={() => {
                            handleInvitationAcceptence();
                          }}
                        >
                          {acceptingInvitation ? <Spinner /> : "Accept"}
                        </PrimaryButton>
                      </div>
                      {invitationError && (
                        <div className={styles.InformationWarning}>
                          {invitationError}
                        </div>
                      )}
                    </div>
                  </div>
                </Modal>
              </div>
            )}
            {displayProjects.length === 0 && isFetched && (
              <div className={styles.NoResourcesMessage}>
                {selectedProjects === "My projects" ? (
                  <>
                    {currentPaginationPage === 1 ? (
                      <div>
                        You haven’t created any projects yet. Click the &nbsp;{" "}
                        <ButtonPlus className={styles.ButtonPlusSmall} /> &nbsp;
                        button to add a project.
                      </div>
                    ) : (
                      <div>
                        This page of you personal projects contains no projects,
                        switch tabs to shared to see Projects &nbsp; Or click{" "}
                        <ButtonPlus className={styles.ButtonPlusSmall} />
                        &nbsp; button to add a project.
                      </div>
                    )}
                  </>
                ) : (
                  <>
                    {currentPaginationPage === 1 ? (
                      <>No project has been shared with you as yet.</>
                    ) : (
                      <>This page contains no shared projects.</>
                    )}
                  </>
                )}
              </div>
            )}
            {!isFetched && !projectsData?.data?.data?.projects?.length === 0 && (
              <div className={styles.NoResourcesMessage}>
                You haven’t created any projects yet. Click the &nbsp;{" "}
                <ButtonPlus className={styles.ButtonPlusSmall} /> &nbsp; button
                to add a project.
              </div>
            )}
            {isError &&(
              <div className={styles.NoResourcesMessage}>
                Oops! Something went wrong! Failed to retrieve Projects.
              </div>
            )}
          </div>
        </div>
      )}
      <div className={styles.FooterRow}>
        {pagination?.pages > 1 &&
          !isRetrieving &&
          isFetched &&
          !openCreateComponent && (
            <div className={styles.PaginationSection}>
              {/* customise pagination for shared and personal projects */}
              <Pagination
                total={pagination?.pages}
                current={currentPaginationPage}
                onPageChange={onPageChange}
              />
            </div>
          )}
        <div>
          Copyright {new Date().getFullYear()} Crane Cloud. All Rights Reserved.
        </div>
      </div>
    </div>
  );
};

UserProjectsPage.propTypes = {
  projects: PropTypes.arrayOf(PropTypes.shape({})),
  clusters: PropTypes.object,
  pagination: PropTypes.shape({}),
  getClustersList: PropTypes.func.isRequired,
  getUserProjects: PropTypes.func.isRequired,
  data: PropTypes.shape({
    id: PropTypes.string,
  }).isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      userID: PropTypes.string,
    }).isRequired,
  }).isRequired,
  isFetched: PropTypes.bool,
  isRetrieving: PropTypes.bool,
};

UserProjectsPage.defaultProps = {
  clusters: [],
  projects: [],
  pagination: {},
  isFetched: false,
  isRetrieving: false,
};

export const mapStateToProps = (state) => {
  const { data } = state.user;

  const { clusters } = state.clustersReducer;
  const { isRetrieving, projects, isFetched, pagination } =
    state.userProjectsReducer;
  const { credits } = state.userCreditsReducer;
  return {
    data,
    isRetrieving,
    projects,
    pagination,
    clusters,
    isFetched,
    credits,
  };
};

const mapDispatchToProps = {
  getUserProjects,
  getClustersList,
  getUserCredits,
};

export default connect(mapStateToProps, mapDispatchToProps)(UserProjectsPage);
