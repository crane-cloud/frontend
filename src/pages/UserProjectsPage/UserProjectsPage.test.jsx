import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect"; // For extended jest-dom assertions
import { Provider } from "react-redux"; // Import your Redux Provider
import configureStore from "redux-mock-store"; // Install the redux-mock-store library
import thunk from "redux-thunk"; // If you're using redux-thunk

import UserProjectsPage from "./";

// Mock your Redux store
const middlewares = [thunk]; // Add any middlewares you're using
const mockStore = configureStore(middlewares);

const initialState = {
  user: {
    data: {
      id: "123",
      username: "testUser",
    },
  },
  clustersReducer: {
    clusters: [],
  },
  userProjectsReducer: {
    isRetrieving: false,
    projects: [],
    pagination: {},
    isFetched: false,
  },
  userCreditsReducer: {
    credits: {},
  },
};

let store;

beforeEach(() => {
  store = mockStore(initialState);
});

it("renders UserProjectsPage component", () => {
  render(
    <Provider store={store}>
      <UserProjectsPage />
    </Provider>
  );

  // Add your testing assertions based on the component's behavior
  expect(screen.getByText("Projects")).toBeInTheDocument();
  // Add more assertions as needed
});

// Add more test cases as needed based on your component's behavior
