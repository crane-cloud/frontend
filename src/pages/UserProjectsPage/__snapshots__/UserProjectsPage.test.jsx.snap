// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the user projects Page component matchs the UserProjectsPage component snapshot 1`] = `ShallowWrapper {}`;
